File: android/src/main/java/com/reactlibrary/RNSpeedTestModule.java

package com.reactlibrary;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.Promise;

public class RNSpeedTestModule extends ReactContextBaseJavaModule {

  private final ReactApplicationContext reactContext;

  public RNSpeedTestModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
  }

  private static final String SPEED_TEST_ERROR = "SPEED_TEST_ERROR";
  @ReactMethod
  public void getSpeed(Promise promise) {
      promise.resolve(200);
  }

  @Override
  public String getName() {
    return "RNSpeedTest";
  }
}

File: android/src/main/java/com/reactlibrary/RNSpeedTestModule.kt
package com.reactlibrary

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.reactlibrary.download.FileDownloadTask
import com.reactlibrary.download.ProgressListener
import java.util.concurrent.TimeUnit

class RNSpeedTestModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    @ReactMethod
    fun getSpeed(promise: Promise) {
        val listener = object : ProgressListener {
            val startSeconds = TimeUnit.NANOSECONDS.toSeconds(System.nanoTime())

            override fun update(bytesRead: Long, contentLength: Long, done: Boolean) {
                if (done) {
                    val endSeconds = TimeUnit.NANOSECONDS.toSeconds(System.nanoTime())
                    val elapsedTime = endSeconds - startSeconds
                    val speed = contentLength / elapsedTime
                    promise.resolve("$speed b/s")
                }
            }
        }
        FileDownloadTask(listener).run()
    }

    override fun getName() = "RNSpeedTest"
}
